### PR TITLE
Chore/compression v1

### PR DESCRIPTION
**Pull request by Tabaie**
# Description

Minor improvements to compression:

* Compressor can decide to bypass compression when appropriate
* One byte at the beginning of the compressed file is dedicated to the compression algorithm version

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

All previous tests pass. No new tests were added.

# How has this been benchmarked?

No benchmarks were run as the change is minimal.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: frontend/cs/scs/api_assertions.go [hunk 1]
```diff
@@ -128,6 +128,33 @@ func (builder *builder) AssertIsBoolean(i1 frontend.Variable) {
 
 }
 
+func (builder *builder) AssertIsCrumb(i1 frontend.Variable) {
+	const errorMsg = "AssertIsCrumb: input is not a crumb"
+	if c, ok := builder.constantValue(i1); ok {
+		bytes := c.Bytes()
```
Comment (gbotrel) [inline frontend/cs/scs/api_assertions.go:134]: I think doing something like 
```
if c.IsUint64() && c.Uint64() <= 3 {return} else {panic(...)} 
```
is more readable

File: frontend/cs/scs/api_assertions.go [hunk 2]
```diff
@@ -128,6 +128,33 @@ func (builder *builder) AssertIsBoolean(i1 frontend.Variable) {
...
+		}
+		for i := 0; i < 4; i++ {
+			if bytes[0] == byte(i) {
+				return
+			}
+		}
+		panic(errorMsg)
+	}
+
+	x := builder.MulAcc(builder.Mul(-3, i1), i1, i1).(expr.Term)
```
Comment (gbotrel) [inline frontend/cs/scs/api_assertions.go:148]: can you comment a bit the logic? i.e. reading the code it's not obvious what is the constraint / if it's sound

File: std/compress/lzss/backref.go [hunk 3]
```diff
@@ -47,22 +47,32 @@ type backref struct {
...
+}
+
 func (b *backref) writeTo(w *bitio.Writer, i int) {
-	w.TryWriteByte(b.bType.delimiter)
-	w.TryWriteBits(uint64(b.length-1), b.bType.nbBitsLength)
-	if b.bType.dictOnly {
-		w.TryWriteBits(uint64(b.address), b.bType.nbBitsAddress)
-	} else {
-		w.TryWriteBits(uint64(i-b.address-1), b.bType.nbBitsAddress)
+	panicIfErr(w.WriteByte(b.bType.delimiter))
```
Comment (gbotrel) [inline std/compress/lzss/backref.go:57]: why the change? The purpose of the `TryWrite...` methods is to make things more readable and test the error afterwards -- probably also has a perf impact over a large stream to do so many branches at each write

File: std/compress/lzss/compress.go [hunk 4]
```diff
@@ -103,7 +103,7 @@ func (compressor *Compressor) Compress(d []byte) (c []byte, err error) {
 
 	// reset output buffer
 	compressor.buf.Reset()
-	compressor.buf.WriteByte(byte(compressor.level))
+	compressor.buf.Write([]byte{0, byte(compressor.level)}) // 0 -> compressor release version
```
Comment (gbotrel) [inline std/compress/lzss/compress.go:106]: can you extract in a `writeMetadata` or `writeHeader`/ corresponding `readMetadata` / `readHeader`?

File: std/compress/lzss/decompress.go [hunk 5]
```diff
@@ -57,59 +65,49 @@ func DecompressGo(data, dict []byte) (d []byte, err error) {
...
-
-	bDict := backref{bType: dictBackRefType}
-	bShort := backref{bType: shortBackRefType}
-	bLong := backref{bType: longBackRefType}
-
+	_ = in.TryReadByte()
 	levelFromData := Level(in.TryReadByte())
-	if levelFromData != NoCompression && levelFromData != level {
-		panic("compression mode mismatch")
+	outLenBits := 16
```
Comment (gbotrel) [inline std/compress/lzss/decompress.go:77]: what is outLen ?

File: std/compress/lzss/compress.go [hunk 6]
```diff
@@ -210,6 +210,14 @@ func (compressor *Compressor) Compress(d []byte) (c []byte, err error) {
 		return nil, err
 	}
 
+	if compressor.buf.Len() >= len(d)+2 {
+		// compression was not worth it
+		out := compressor.buf.Bytes()
+		out[1] = byte(NoCompression)
```
Comment (gbotrel) [inline std/compress/lzss/compress.go:216]: use a clean `writeHeader` or `writeMetadata` func so the logic to read / write that is not dispersed

File: std/compress/stream.go [hunk 7]
```diff
@@ -31,12 +31,21 @@ func (s *Stream) At(i int) int {
 	return s.D[i]
 }
 
-func NewStreamFromBytes(in []byte) Stream {
-	d := make([]int, len(in))
-	for i := range in {
-		d[i] = int(in[i])
+func panicIfErr(err error) {
+	if err != nil {
```
Comment (gbotrel) [inline std/compress/stream.go:35]: I would rather panics be real panics; if they are actually reachable due to a malformed input or something like that, it's a potential (unlikely but still ^^) ddos vector

File: frontend/cs/scs/api_assertions.go (hunk 2 above)
Comment (gbotrel) [reply frontend/cs/scs/api_assertions.go:148]: + also interested to know the justification of this API (i.e clearly say; using the `std/math/bits` package generates X constraints while this generates Y constraint for R1CS, Z for SCS)

**Review by gbotrel — CHANGES_REQUESTED** [review]

File: std/compress/lzss/backref.go (hunk 3 above)
Comment (Tabaie) [reply std/compress/lzss/backref.go:57]: The errors overwrite each other so if we encounter an error in one of the middle writes (don't know how likely that is; to fail on one write and succeed on the following ones) we won't know that there ever was an error.

**Review by gbotrel — APPROVED** [review]
